src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs

```diff
@@ -355,7 +355,7 @@ private void BoundsCheck(in SequencePosition position)
                 // Storing this in a local since it is used twice within InRange()
                 ulong startRange = (ulong)(((ReadOnlySequenceSegment<T>)startObject!).RunningIndex + startIndex);
                 if (!InRange(
-                    (ulong)(((ReadOnlySequenceSegment<T>)position.GetObject()!).RunningIndex + sliceStartIndex),
+                    (ulong)((position.GetObject() != null ? ((ReadOnlySequenceSegment<T>)position.GetObject()!).RunningIndex : 0) + sliceStartIndex),
```

Comment:
This seems counter to the annotated bang (!) on `position.GetObject()`. @safern - what should be the correct annotation here? Should we remove the `!`?

nit: We may want to extract this out in a local to help with readability.

Reply:
The compiler will not do deep data flow analysis here to track that you already checked for null on the call of `GetObject()` and that you're calling it again when it is not null. So unfortunately the `!` is still needed here. I just validated that: https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKFwYgDsBXAG1ImFLgAI5DLrcUBmG9GgYRoG9caB7NgEtCMGgFkAFABEaABwD2UYTGGLCASl79BASBQB2BctXrCAOgDicGAHlgAKzgBjGFO0BCALw0S5GgB+GikpAC1NJRU1DWtbB2c3D00LAElCABM4AA8aEBoABgBuXQEAX1wKvBxWdjQaOT4cPVKaRSdXGGCAfXbE8V9/UhKcQRpW2r7O4Jt7DqTtbwA+Gl75mBGq5jYOMJ1m2tFxdKzcnhoAc1simihrmjKaXwxNoA

If perf is not affected and adding an extra local doesn't hurt, that would be the only way to remove the `!`, but if that hurts the produced IL we should leave as is.

cc: @cston I guess this behavior is by design and not something the compiler can infer?

Reply:
@safern, yes, this is by design - the compiler tracks the nullability of locals, parameters, and fields and properties of those only.